clientgenv2/source_generator.go

```diff
@@ -76,6 +69,97 @@ func (rs ResponseFieldList) IsStructType() bool {
 	return len(rs) > 0 && !rs.IsFragment()
 }
 
+type StructGenerator struct {
+	currentResponseFieldList *ResponseFieldList
+	preMergedStructSources   []*StructSource
+	postMergedStructSources  []*StructSource
+}
+
+func NewStructGenerator(responseFieldList ResponseFieldList) *StructGenerator {
+	currentFields := make(ResponseFieldList, 0)
+	fragmentChildrenFields := make(ResponseFieldList, 0)
+	for _, field := range responseFieldList {
+		if field.IsFragmentSpread {
+			fragmentChildrenFields = append(fragmentChildrenFields, field.ResponseFields...)
+		} else {
+			currentFields = append(currentFields, field)
+		}
+	}
+
+	preMergedStructSources := make([]*StructSource, 0)
+	postMergedStructSources := make([]*StructSource, 0)
+
+	for _, field := range responseFieldList {
+		if field.IsFragmentSpread {
+			preMergedStructSources = append(preMergedStructSources, &StructSource{
+				Name: field.FieldTypeString(),
+				Type: field.ResponseFields.StructType(),
+			})
+		}
+	}
+
+	currentFields = mergeFieldsRecursively(currentFields, fragmentChildrenFields, &preMergedStructSources, &postMergedStructSources)
+	return &StructGenerator{
+		currentResponseFieldList: &currentFields,
+		preMergedStructSources:   preMergedStructSources,
+		postMergedStructSources:  postMergedStructSources,
+	}
+}
+
+func mergeFieldsRecursively(targetFields ResponseFieldList, sourceFields ResponseFieldList, preMerged, postMerged *[]*StructSource) ResponseFieldList {
+	for _, sourceField := range sourceFields {
+		sameNameFieldFlag := false
+		for _, targetField := range targetFields {
+			if sourceField.Name == targetField.Name && !targetField.ResponseFields.IsBasicType() {
+				*preMerged = append(*preMerged, &StructSource{
+					Name: sourceField.FieldTypeString(),
+					Type: sourceField.ResponseFields.StructType(),
+				})
+				*preMerged = append(*preMerged, &StructSource{
+					Name: targetField.FieldTypeString(),
+					Type: targetField.ResponseFields.StructType(),
+				})
+
+				targetField.ResponseFields = mergeFieldsRecursively(targetField.ResponseFields, sourceField.ResponseFields, preMerged, postMerged)
+				*postMerged = append(*postMerged, &StructSource{
+					Name: targetField.FieldTypeString(),
+					Type: targetField.ResponseFields.StructType(),
+				})
+				sameNameFieldFlag = true
+				break
+			}
+		}
+		// if there is no same name field, append it
+		if !sameNameFieldFlag {
+			targetFields = append(targetFields, sourceField)
```

Comment:
It’s not ideal to cause side effects on arguments, so it would be better to define a type for the return value.

Reply:
okey, I fixed it